Introduce AllowTestOnlyIPC in order to block "ForTesting" IPC endpoints

gavin-apple opened the pull request:
#### 106f2630430967aba20df89a91f093284a7f4362
<pre>
Introduce AllowTestOnlyIPC in order to block &quot;ForTesting&quot; IPC endpoints
<a href="https://bugs.webkit.org/show_bug.cgi?id=252829">https://bugs.webkit.org/show_bug.cgi?id=252829</a>
rdar://105837347

Reviewed by Alex Christensen.

Starts to block unneeded &quot;ForTesting&quot; IPC endpoints by adding a default
deny to those messages and then allowing them only when a preference is
specified. The new preference exposed is AllowTestOnlyIPC.

* LayoutTests/TestExpectations:
* LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption.html:
* LayoutTests/fast/mediastream/video-created-while-interrupted.html:
* LayoutTests/ipc/no-test-only-ipc-expected-crash.html: Added.
* LayoutTests/ipc/test-only-ipc-allowed-expected.txt: Added.
* LayoutTests/ipc/test-only-ipc-allowed.html: Added.
* LayoutTests/media/audioSession/audioSessionState.html:
* LayoutTests/media/audioSession/ios/audioSessionState-getUserMedia.html:
* LayoutTests/platform/ios/mediastream/getUserMedia-override-audio-session-interruption.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::setMediaOverridesForTesting):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::allowTestOnlyIPC const):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::triggerBeginInterruptionForTesting):
(WebKit::RemoteAudioSessionProxy::triggerEndInterruptionForTesting):
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
(WebKit::GPUProcessConnectionParameters::encode const): Deleted.
(WebKit::GPUProcessConnectionParameters::decode): Deleted.
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in: Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _allowTestOnlyIPC]):
(-[WKWebViewConfiguration _setAllowTestOnlyIPC:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createGPUProcessConnection):
(WebKit::WebProcessPool::createWebPage):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::allowTestOnlyIPC const):
(WebKit::WebProcessProxy::setAllowTestOnlyIPC):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::allowTestOnlyIPC const):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::platformCreateWebView):

Canonical link: <a href="https://commits.webkit.org/260935@main">https://commits.webkit.org/260935@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ece1e3a7dcc72fe8b8aa8a8f41eb8f2ef2e327d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109989 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19090 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42672 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1433 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119036 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113941 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20555 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10283 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102247 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115736 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43520 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30176 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85343 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11792 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31516 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12415 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17783 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51127 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7581 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14210 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -181,6 +181,7 @@ @implementation WKWebViewConfiguration {
 #if ENABLE(APP_HIGHLIGHTS)
     BOOL _appHighlightsEnabled;
 #endif
+    BOOL _allowTestOnlyIPC;
```
This should be on API::PageConfiguration instead of adding to the ObjC ivars.

achristensen07 commented on LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption.html:
```diff
@@ -1,4 +1,4 @@
-<!DOCTYPE html>
+<!DOCTYPE html> <!-- webkit-test-runner [ AllowTestOnlyIPC=true ] -->
```
This could just be on for all WebKitTestRunner WKWebViews.  This will get annoying really quickly.

achristensen07 commented on Source/WebKit/Shared/GPUProcessConnectionParameters.h:
```diff
@@ -40,67 +40,16 @@ struct GPUProcessConnectionParameters {
 };
-
-} // namespace WebKit
+    
```
whitespace

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -891,6 +891,9 @@ void WebPageProxy::launchProcess(const RegistrableDomain& registrableDomain, Pro
         m_process->setIgnoreInvalidMessageForTesting();
 #endif
+    
+    if (m_preferences->allowTestOnlyIPC())
```
Get this from the configuration instead of preferences.  Then remove it from the yaml

gavin-apple committed:
Introduce AllowTestOnlyIPC in order to block "ForTesting" IPC endpoints
https://bugs.webkit.org/show_bug.cgi?id=252829
rdar://105837347

Reviewed by Alex Christensen.

Starts to block unneeded "ForTesting" IPC endpoints by adding a default
deny to those messages and then allowing them only when a preference is
specified. The new preference exposed is AllowTestOnlyIPC.

* LayoutTests/TestExpectations:
* LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption.html:
* LayoutTests/fast/mediastream/video-created-while-interrupted.html:
* LayoutTests/ipc/no-test-only-ipc-expected-crash.html: Added.
* LayoutTests/ipc/test-only-ipc-allowed-expected.txt: Added.
* LayoutTests/ipc/test-only-ipc-allowed.html: Added.
* LayoutTests/media/audioSession/audioSessionState.html:
* LayoutTests/media/audioSession/ios/audioSessionState-getUserMedia.html:
* LayoutTests/platform/ios/mediastream/getUserMedia-override-audio-session-interruption.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::setMediaOverridesForTesting):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::allowTestOnlyIPC const):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::triggerBeginInterruptionForTesting):
(WebKit::RemoteAudioSessionProxy::triggerEndInterruptionForTesting):
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
(WebKit::GPUProcessConnectionParameters::encode const): Deleted.
(WebKit::GPUProcessConnectionParameters::decode): Deleted.
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in: Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _allowTestOnlyIPC]):
(-[WKWebViewConfiguration _setAllowTestOnlyIPC:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createGPUProcessConnection):
(WebKit::WebProcessPool::createWebPage):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::allowTestOnlyIPC const):
(WebKit::WebProcessProxy::setAllowTestOnlyIPC):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::allowTestOnlyIPC const):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::platformCreateWebView):

Canonical link: https://commits.webkit.org/260935@main
